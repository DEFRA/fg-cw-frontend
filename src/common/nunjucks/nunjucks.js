import Vision from "@hapi/vision";
import Nunjucks from "nunjucks";

import { glob } from "node:fs/promises";
import { config } from "../config.js";
import { filters } from "../format/format-filters.js";
import { context } from "./context/context.js";
import { formatDate } from "./filters/format-date.js";

const njkPaths = [
  "node_modules/govuk-frontend/dist/",
  ...(await Array.fromAsync(
    glob(["**/views", "**/*(layouts|components|pages|partials)/"], {
      exclude: ["node_modules", "*.*"],
    }),
  )),
];

const viewPaths = await Array.fromAsync(
  glob(["**/views"], {
    exclude: ["node_modules", "*.*"],
  }),
);

export const environment = Nunjucks.configure(njkPaths, {
  autoescape: true,
  throwOnUndefined: false,
  trimBlocks: true,
  lstripBlocks: true,
  watch: config.get("nunjucks.watch"),
  noCache: config.get("nunjucks.noCache"),
  filter: {
    formatDate,
  },
});

environment.addFilter("formatDate", formatDate);
environment.addFilter("fixed", filters.fixed);

export const nunjucks = {
  plugin: Vision,
  options: {
    engines: {
      njk: {
        compile(src, options) {
          const template = Nunjucks.compile(src, options.environment);
          return (ctx) => template.render(ctx);
        },
      },
    },
    compileOptions: {
      environment,
    },
    relativeTo: config.get("root"),
    path: viewPaths,
    isCached: config.get("isProduction"),
    context,
  },
};
