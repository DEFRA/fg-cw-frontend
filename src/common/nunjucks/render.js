import camelCase from "lodash/camelCase.js";
import { environment } from "./nunjucks.js";

/**
 * @param {string} name
 * @param {object} params
 * @param {string} [block]
 */
export const render = (name, params, block) => {
  const macro = {
    path: `${name}/macro.njk`,
    name: camelCase(name.split("-")),
    params: JSON.stringify(params, null, 2),
  };

  const callComponent = block
    ? `{%- call ${macro.name}(${macro.params}) -%}${block}{%- endcall -%}`
    : `{{- ${macro.name}(${macro.params}) -}}`;

  return environment.renderString(
    `{%- from "${macro.path}" import ${macro.name} -%}
    ${callComponent}
    `,
    {},
  );
};
