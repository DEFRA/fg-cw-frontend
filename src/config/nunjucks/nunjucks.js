import { fileURLToPath } from 'node:url'
import path from 'path'
import nunjucks from 'nunjucks'
import hapiVision from '@hapi/vision'

import { config } from '../config.js'
import { context } from './context/context.js'
import * as filters from './filters/filters.js'
import * as globals from './globals.js'

const dirname = path.dirname(fileURLToPath(import.meta.url))
const nunjucksEnvironment = nunjucks.configure(
  [
    'node_modules/govuk-frontend/dist/',
    path.resolve(dirname, '../../server/common/templates'),
    path.resolve(dirname, '../../server/common/components'),
    path.resolve(dirname, '../../case-management/views'),
    path.resolve(dirname, '../../case-management/views/layouts'),
    path.resolve(dirname, '../../case-management/views/components'),
    path.resolve(dirname, '../../case-management/views/pages'),
    path.resolve(dirname, '../../case-management/views/partials'),
    path.resolve(dirname, '../../common/presentation')
  ],
  {
    autoescape: true,
    throwOnUndefined: false,
    trimBlocks: true,
    lstripBlocks: true,
    watch: config.get('nunjucks.watch'),
    noCache: config.get('nunjucks.noCache')
  }
)

export const nunjucksConfig = {
  plugin: hapiVision,
  options: {
    engines: {
      njk: {
        compile(src, options) {
          const template = nunjucks.compile(src, options.environment)
          return (ctx) => template.render(ctx)
        }
      }
    },
    compileOptions: {
      environment: nunjucksEnvironment
    },
    relativeTo: path.resolve(dirname, '../..'),
    path: ['server', 'case-management/views', 'common/presentation'],
    isCached: config.get('isProduction'),
    context
  }
}

Object.entries(globals).forEach(([name, global]) => {
  nunjucksEnvironment.addGlobal(name, global)
})

Object.entries(filters).forEach(([name, filter]) => {
  nunjucksEnvironment.addFilter(name, filter)
})
