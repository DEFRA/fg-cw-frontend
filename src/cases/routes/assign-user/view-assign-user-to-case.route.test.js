import hapi from "@hapi/hapi";
import { load } from "cheerio";
import { afterAll, beforeAll, describe, expect, it, vi } from "vitest";
import { findAllUsersUseCase } from "../../../auth/use-cases/find-all-users.use-case.js";
import { nunjucks } from "../../../common/nunjucks/nunjucks.js";
import { findCaseByIdUseCase } from "../../use-cases/find-case-by-id.use-case.js";
import { viewAssignUserToCaseRoute } from "./view-assign-user-to-case.route.js";

vi.mock("../../use-cases/find-case-by-id.use-case.js");
vi.mock("../../../auth/use-cases/find-all-users.use-case.js");

describe("viewAssignUserToCaseRoute", () => {
  let server;

  beforeAll(async () => {
    server = hapi.server();
    server.route(viewAssignUserToCaseRoute);
    await server.register([nunjucks]);

    await server.initialize();
  });

  afterAll(async () => {
    await server.stop();
  });

  it("returns a case and list of valid users", async () => {
    findCaseByIdUseCase.mockResolvedValue(mockCase);
    findAllUsersUseCase.mockResolvedValue(mockUsers);

    const { statusCode, result } = await server.inject({
      method: "GET",
      url: "/cases/assign-user?caseId=6870ee690cdf25de1301a300",
      auth: {
        credentials: { token: "mock-token" },
        strategy: "session",
      },
    });

    expect(statusCode).toEqual(200);

    const $ = load(result);
    const view = $("#main-content").html();

    expect(view).toMatchSnapshot();
  });

  it("redirects back to cases list if no caseId supplied", async () => {
    const { statusCode, headers } = await server.inject({
      method: "GET",
      url: "/cases/assign-user",
      auth: {
        credentials: { token: "mock-token" },
        strategy: "session",
      },
    });

    expect(statusCode).toEqual(302);
    expect(headers.location).toEqual("/cases");
  });
});

const mockCase = {
  _id: "case-123",
  payload: {
    clientRef: "CLIENT-001",
    code: "CODE-001",
  },
  requiredRoles: {
    allOf: ["admin", "manager"],
    anyOf: ["reviewer", "approver"],
  },
  assignedUser: { id: "user-1" },
};

const mockUsers = [
  {
    id: "user-1",
    name: "John Doe",
    appRoles: ["admin", "manager", "reviewer"],
  },
  {
    id: "user-2",
    name: "Jane Smith",
    appRoles: ["admin", "manager", "approver"],
  },
];
