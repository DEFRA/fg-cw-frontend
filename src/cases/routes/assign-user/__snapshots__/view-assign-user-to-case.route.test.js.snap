// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`viewAssignUserToCaseRoute > returns a case and list of valid users 1`] = `
"
  <div class="govuk-grid-row app-heading" data-testid="app-heading">
  <div class="govuk-grid-column-two-thirds">
    <h1 class="govuk-heading-xl govuk-!-margin-bottom-2" data-testid="app-heading-title">
      Assign
    </h1>

  </div>
</div>


  <div class="govuk-body" data-testid="app-page-body">
    <form action="/cases/assign-user" method="post">
      <input type="hidden" name="caseId" value="case-123">

      <div class="govuk-form-group">
  <label class="govuk-label" for="assignedUserId">
    Case Worker
  </label>
  <select class="govuk-select" id="assignedUserId" name="assignedUserId">
    <option value="">Select a user</option>
    <option value="user-1" selected="">John Doe</option>
    <option value="user-2">Jane Smith</option>
  </select>
</div>


      <div class="govuk-form-group">
  <h1 class="govuk-label-wrapper">
    <label class="govuk-label govuk-label--l" for="notes">
      Notes (optional)
    </label>
  </h1>
  <textarea class="govuk-textarea" id="notes" name="notes" rows="5"></textarea>
</div>


      <button type="submit" class="govuk-button" data-module="govuk-button">
  Assign
</button>
    </form>
  </div>
        "
`;
